File: app/src/main/java/com/solverlabs/droid/rugl/GameActivity.java
package com.solverlabs.droid.rugl;

import android.app.ProgressDialog;
import android.content.Context;
import android.text.InputFilter;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.solverlabs.droid.rugl.res.ResourceLoader;
import com.solverlabs.worldcraft.BlockView;
import com.solverlabs.worldcraft.Enemy;
import com.solverlabs.worldcraft.GameMode;
import com.solverlabs.worldcraft.R;
import com.solverlabs.worldcraft.SoundManager;
import com.solverlabs.worldcraft.activity.CommonActivity;
import com.solverlabs.worldcraft.factories.DescriptionFactory;
import com.solverlabs.worldcraft.multiplayer.Multiplayer;
import com.solverlabs.worldcraft.multiplayer.dialogs.PopupDialog;
import com.solverlabs.worldcraft.multiplayer.util.TextUtils;
import com.solverlabs.worldcraft.ui.ChatBox;
import com.solverlabs.worldcraft.util.KeyboardUtils;

import java.util.ArrayList;
import java.util.Set;
import java.util.TreeSet;

/**
 * Handy activity that can be simply subclassed. Just remember to call
 * {@link #start(Game, String)} in your
 * {@link #onCreate(android.os.Bundle)} or nothing will happen.
 * Handles starting the {@link ResourceLoader} and key input
 */
public abstract class GameActivity extends CommonActivity implements Runnable {
    private static final String CHAT_COMMAND_HOME = "/home";
    /**
     * The {@link Game}
     */
    protected Game game;
    protected ProgressDialog loadingDialog;
    protected AlertDialog readOnlyMapNotificationDialog;
    private GameView gameView;

    /**
     * Call this in your {@link #onCreate(android.os.Bundle)}
     * implementation
     *
     * @param game
     */
    public void start(Game game) {
        this.game = game;
        ResourceLoader.start(getResources());
        setContentView(R.layout.activity_game);
        this.gameView = findViewById(R.id.gameViewWithoutBanner);
        this.gameView.init(game);
        Thread reportAbuseCatcher = new Thread(this);
        reportAbuseCatcher.start();
    }

    /**
     * Displays a short message to the user
     *
     * @param message
     * @param longShow <code>true</code> for {@link Toast#LENGTH_LONG},
     *                 <code>false</code> for {@link Toast#LENGTH_SHORT}
     */
    public void showToast(final String message, final boolean longShow) {
        runOnUiThread(() -> {
            Toast t = Toast.makeText(getApplicationContext(), message, longShow ? Toast.LENGTH_LONG : Toast.LENGTH_SHORT);
            t.show();
        });
    }

    public void showGameMenuDialog() {
        final View view = View.inflate(this, R.layout.menulayout, null);
        Button quitButton = view.findViewById(R.id.quitButton);
        Button backButton = view.findViewById(R.id.backButton);
        Button playersList = view.findViewById(R.id.playerListButton);

        final MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(this);
        builder.setView(view);
        builder.setTitle("Game menu");
        final AlertDialog dialog = builder.create();
        if (GameMode.isMultiplayerMode()) {
            playersList.setVisibility(View.VISIBLE);
        } else {
            playersList.setVisibility(View.GONE);
        }
        quitButton.setOnClickListener(v -> {
            if (GameMode.isMultiplayerMode()) {
                showLikeDialog();
            } else {
                completeCurrentPhase(true);
            }
            SoundManager.stopAllSounds();
            dialog.dismiss();
        });
        backButton.setOnClickListener(v -> dialog.dismiss());
        playersList.setOnClickListener(v -> {
            showPlayerList();
            dialog.dismiss();
        });
        dialog.show();
    }

    public void showPlayerList() {
        final View view = View.inflate(this, R.layout.playerlist, null);

        final MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(this);
        builder.setView(view);
        ArrayList<String> list = new ArrayList<>();
        builder.setTitle("Player list       Room name:  " + Multiplayer.instance.roomName);
        list.add(Multiplayer.instance.playerName + "   (you)");
        Set<Enemy> sortedEnemies = new TreeSet<>(Multiplayer.getEnemiesCopy());
        for (Enemy enemy : sortedEnemies) {
            list.add(enemy.name);
        }
        final AlertDialog dialog = builder.create();

        ArrayAdapter<String> dataAdapter = new ArrayAdapter<>(this, R.layout.custom_list_content, R.id.list_content, list);
        ListView playerList = view.findViewById(R.id.playerListView);
        playerList.setAdapter(dataAdapter);
        Button cancelButton = view.findViewById(R.id.cancel);
        cancelButton.setOnClickListener(v -> dialog.dismiss());
        dialog.show();
    }

    private void showLoadingDialog(final String message) {
        try {
            runOnUiThread(() -> {
                if (!isFinishing()) {
                    loadingDialog = ProgressDialog.show(GameActivity.this, DescriptionFactory.emptyText, message, true);
                    loadingDialog.setCancelable(false);
                    loadingDialog.show();
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void dismissLoadingDialog() {
        try {
            runOnUiThread(() -> {
                if (loadingDialog != null && !loadingDialog.isShowing()) {
                    loadingDialog.dismiss();
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void saveWorld(final String worldName) {
        new Thread() {
            @Override
            public void run() {
                try {
                    gameView.game.getBlockView().saveWorld(worldName);
                } catch (Throwable th) {
                    th.printStackTrace();
                }
                try {
                    completeCurrentPhase(true);
                } catch (Throwable th) {
                    th.printStackTrace();
                }
            }
        }.start();
    }

    public void completeCurrentPhase(final boolean needSaveWorld) {
        new Thread() {
            @Override
            public void run() {
                try {
                    gameView.game.getBlockView().complete(needSaveWorld);
                } catch (Throwable th) {
                    th.printStackTrace();
                }
                try {
                    Multiplayer.instance.shutdown();
                    dismissLoadingDialog();
                } catch (Throwable th) {
                    th.printStackTrace();
                }
            }
        }.start();
    }

    public void showSaveWorldDialogOnConnectionLost() {
        showSaveWorldDialog(R.string.connection_lost_would_you_like_to_save_world);
    }

    public void showSaveWorldDialog() {
        showSaveWorldDialog(R.string.would_you_like_to_save_world);
    }

    public void showReadOnlyRoomModificationDialog() {
        showSaveWorldDialog(R.string.modification_read_only_world_warning, 17039360, false);
    }

    public void showSaveWorldDialog(int stringId) {
        showSaveWorldDialog(stringId, R.string.dont_save, true);
    }

    public void showSaveWorldDialog(final int stringId, final int noButtonResourseId, final boolean completePhaseOnNoClick) {
        runOnUiThread(() -> {
            if (readOnlyMapNotificationDialog == null || !readOnlyMapNotificationDialog.isShowing()) {
                final EditText input = new EditText(GameActivity.this);
                input.setHint(R.string.world_name);
                input.setText(Multiplayer.instance.roomName);
                KeyboardUtils.hideKeyboardOnEnter(GameActivity.this, input);
                final MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(this);
                builder.setTitle(R.string.save_world);
                builder.setMessage(stringId);
                builder.setView(input);
                builder.setCancelable(false);
                builder.setPositiveButton(R.string.save_world, (dialog, id) -> {
                    String worldName = String.valueOf(input.getText());
                    if (!DescriptionFactory.emptyText.equals(worldName) && !"null".equals(worldName)) {
                        saveWorld(worldName);
                        hideKeyBoard(input);
                        return;
                    }
                    Toast.makeText(GameActivity.this, R.string.wrong_world_name, Toast.LENGTH_LONG).show();
                });
                builder.setNeutralButton(noButtonResourseId, (dialog, id) -> {
                    if (completePhaseOnNoClick) {
                        completeCurrentPhase(false);
                    }
                });
                readOnlyMapNotificationDialog = builder.create();
                readOnlyMapNotificationDialog.getWindow().setSoftInputMode(2);
                readOnlyMapNotificationDialog.show();
            }
        });
    }

    public void showLikeDialog() {
        final MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(this);
        builder.setTitle(R.string.like_it);
        builder.setMessage(R.string.did_you_like_this_world);
        builder.setCancelable(false);
        builder.setPositiveButton(R.string.like, (dialog, id) -> {
            Multiplayer.instance.likeWorld();
            dialog.dismiss();
            showSaveWorldDialog();
        });
        builder.setNegativeButton(R.string.dislike, (dialog, id) -> {
            Multiplayer.instance.dislikeWorld();
            dialog.dismiss();
            showSaveWorldDialog();
        });
        builder.show();
    }

    public void sendChatMessage(@NonNull final EditText msg) {
        BlockView phase;
        final String messageText = String.valueOf(msg.getText());
        if (CHAT_COMMAND_HOME.equals(messageText)) {
            if (this.game != null && this.game.getBlockView() != null && (phase = this.game.getBlockView()) != null && (phase instanceof BlockView)) {
                phase.resetPlayerLocation();
                return;
            }
            return;
        }
        new Thread() {
            @Override
            public void run() {
                if (Multiplayer.instance.gameClient != null && TextUtils.isEditTextMessageEmpty(msg)) {
                    Multiplayer.instance.gameClient.chat(messageText);
                }
            }
        }.start();
    }

    public void showChatDialog() {
        final MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(this);
        final EditText msg = new EditText(this);
        msg.setFilters(new InputFilter[]{new InputFilter.LengthFilter(ChatBox.getMaxChatMessageLength())});
        builder.setTitle(R.string.your_message).setView(msg).setPositiveButton(android.R.string.ok, (dialog, id) -> sendChatMessage(msg)).setNegativeButton(R.string.cancel, (dialog, id) -> dialog.dismiss());
        final AlertDialog alert = builder.create();
        msg.setOnEditorActionListener((v, actionId, event) -> {
            if (event == null || event.getKeyCode() != 66) {
                return false;
            }
            sendChatMessage(msg);
            alert.dismiss();
            return true;
        });
        alert.show();
    }

    @Override
    public void onPause() {
        super.onPause();
        dismissLoadingDialog();
        if (this.gameView != null) {
            this.gameView.onPause();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if (this.gameView != null) {
            this.gameView.onResume();
        }
        if (this.game != null) {
            this.game.resetTouches();
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, @NonNull KeyEvent event) {
        BlockView p;
        if (event.getRepeatCount() == 0 && this.gameView != null && this.gameView.game != null && (p = this.gameView.game.getBlockView()) != null) {
            if (keyCode == 4 && (p instanceof BlockView)) {
                showGameMenuDialog();
            } else {
                p.onKeyDown(keyCode, event);
            }
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        BlockView p;
        if (this.gameView != null && this.gameView.game != null && (p = this.gameView.game.getBlockView()) != null) {
            p.onKeyUp(keyCode, event);
            return true;
        }
        return true;
    }

    public void hideKeyBoard(@NonNull View v) {
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
    }

    @Override
    public void finish() {
        super.finish();
    }

    @Override
    public void run() {
        while (GameMode.isMultiplayerMode()) {
            String message = Multiplayer.pollPopupMessage();
            if (message != null) {
                PopupDialog.showInUiThread(R.string.warning, message, this);
            }
            try {
                Thread.sleep(1000L);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/com/mcal/droid/rugl/GameActivity.kt
package com.mcal.droid.rugl

import android.app.ProgressDialog
import android.content.DialogInterface
import android.text.InputFilter
import android.text.InputFilter.LengthFilter
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.appcompat.app.AlertDialog
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.mcal.droid.rugl.res.ResourceLoader
import com.mcal.worldcraft.Enemy
import com.mcal.worldcraft.GameMode.isMultiplayerMode
import com.mcal.worldcraft.R
import com.mcal.worldcraft.SoundManager
import com.mcal.worldcraft.activity.BaseActivity
import com.mcal.worldcraft.databinding.MenulayoutBinding
import com.mcal.worldcraft.factories.DescriptionFactory
import com.mcal.worldcraft.multiplayer.Multiplayer
import com.mcal.worldcraft.multiplayer.dialogs.PopupDialog
import com.mcal.worldcraft.ui.ChatBox
import com.mcal.worldcraft.utils.KeyboardUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.util.*

/**
 * Handy activity that can be simply subclassed. Just remember to call
 * [.start] in your
 * [.onCreate] or nothing will happen.
 * Handles starting the [ResourceLoader] and key input
 */
abstract class GameActivity : BaseActivity() {
    /**
     * The [Game]
     */
    protected var game: Game? = null
    private var loadingDialog: ProgressDialog? = null
    private var readOnlyMapNotificationDialog: AlertDialog? = null
    private var gameView: GameView? = null

    init {
        CoroutineScope(Dispatchers.IO).launch {
            while (isMultiplayerMode) {
                Multiplayer.pollPopupMessage()?.let { message ->
                    PopupDialog.showInUiThread(R.string.warning, message, this@GameActivity)
                }
                delay(1000L)
            }
        }
    }

    /**
     * Call this in your [.onCreate]
     * implementation
     *
     * @param game
     */
    fun start(game: Game) {
        this.game = game
        ResourceLoader.start(resources)
        setContentView(R.layout.activity_game)
        gameView = findViewById<GameView>(R.id.gameViewWithoutBanner).apply {
            init(game)
        }
    }

    /**
     * Displays a short message to the user
     *
     * @param message
     * @param longShow `true` for [Toast.LENGTH_LONG],
     * `false` for [Toast.LENGTH_SHORT]
     */
    fun showToast(message: String?, longShow: Boolean) {
        runOnUiThread {
            Toast.makeText(
                applicationContext,
                message,
                if (longShow) Toast.LENGTH_LONG else Toast.LENGTH_SHORT
            ).show()
        }
    }

    fun showGameMenuDialog() {
        val view = MenulayoutBinding.inflate(LayoutInflater.from(this))
        val playersList = view.playerListButton
        val dialog = MaterialAlertDialogBuilder(this).apply {
            setTitle("Game menu")
            setView(view.root)
        }.create()
        if (isMultiplayerMode) {
            playersList.visibility = View.VISIBLE
        } else {
            playersList.visibility = View.GONE
        }
        view.quitButton.setOnClickListener {
            if (isMultiplayerMode) {
                showLikeDialog()
            } else {
                completeCurrentPhase(true)
            }
            SoundManager.stopAllSounds()
            dialog.dismiss()
        }
        view.backButton.setOnClickListener { dialog.dismiss() }
        playersList.setOnClickListener {
            showPlayerList()
            dialog.dismiss()
        }
        dialog.show()
    }

    private fun showPlayerList() {
        val view = View.inflate(this, R.layout.playerlist, null)
        val builder = MaterialAlertDialogBuilder(this)
        builder.setView(view)
        builder.setTitle("Player list       Room name:  " + Multiplayer.instance.roomName)
        val list = ArrayList<String>()
        list.add(Multiplayer.instance.playerName + "   (you)")
        val sortedEnemies: Set<Enemy> = TreeSet(Multiplayer.getEnemiesCopy())
        for (enemy in sortedEnemies) {
            list.add(enemy.name)
        }
        val dialog = builder.create()
        view.findViewById<ListView>(R.id.playerListView).apply {
            adapter = ArrayAdapter(
                this@GameActivity,
                R.layout.custom_list_content,
                R.id.list_content,
                list
            )
        }
        val cancelButton = view.findViewById<Button>(R.id.cancel)
        cancelButton.setOnClickListener { dialog.dismiss() }
        dialog.show()
    }

    private fun showLoadingDialog(message: String) {
        try {
            runOnUiThread {
                if (!isFinishing) {
                    loadingDialog = ProgressDialog.show(
                        this@GameActivity,
                        DescriptionFactory.emptyText,
                        message,
                        true
                    ).apply {
                        setCancelable(false)
                        show()
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun dismissLoadingDialog() {
        try {
            runOnUiThread {
                val dialog = loadingDialog
                if (dialog != null && !dialog.isShowing) {
                    dialog.dismiss()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun saveWorld(worldName: String?) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                gameView?.game?.blockView?.saveWorld(worldName)
            } catch (th: Throwable) {
                th.printStackTrace()
            }
            try {
                completeCurrentPhase(true)
            } catch (th: Throwable) {
                th.printStackTrace()
            }
        }
    }

    fun completeCurrentPhase(needSaveWorld: Boolean) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                gameView?.game?.blockView?.complete(needSaveWorld)
            } catch (th: Throwable) {
                th.printStackTrace()
            }
            try {
                Multiplayer.instance.shutdown()
                dismissLoadingDialog()
            } catch (th: Throwable) {
                th.printStackTrace()
            }
        }
    }

    fun showSaveWorldDialogOnConnectionLost() {
        showSaveWorldDialog(R.string.connection_lost_would_you_like_to_save_world)
    }

    fun showReadOnlyRoomModificationDialog() {
        showSaveWorldDialog(R.string.modification_read_only_world_warning, 17039360, false)
    }

    @JvmOverloads
    fun showSaveWorldDialog(
        stringId: Int = R.string.would_you_like_to_save_world,
        noButtonResourseId: Int = R.string.dont_save,
        completePhaseOnNoClick: Boolean = true
    ) {
        runOnUiThread {
            var dialog = readOnlyMapNotificationDialog
            if (dialog == null || !dialog.isShowing) {
                val input = EditText(this@GameActivity)
                input.setHint(R.string.world_name)
                input.setText(Multiplayer.instance.roomName)
                KeyboardUtils.hideKeyboardOnEnter(this@GameActivity, input)
                val builder = MaterialAlertDialogBuilder(this)
                builder.setTitle(R.string.save_world)
                builder.setMessage(stringId)
                builder.setView(input)
                builder.setCancelable(false)
                builder.setPositiveButton(R.string.save_world) { _: DialogInterface?, _: Int ->
                    val worldName = input.text.toString()
                    if (DescriptionFactory.emptyText != worldName && "null" != worldName) {
                        saveWorld(worldName)
                        hideKeyBoard(input)
                        return@setPositiveButton
                    }
                    Toast.makeText(this@GameActivity, R.string.wrong_world_name, Toast.LENGTH_LONG)
                        .show()
                }
                builder.setNeutralButton(noButtonResourseId) { _: DialogInterface?, _: Int ->
                    if (completePhaseOnNoClick) {
                        completeCurrentPhase(false)
                    }
                }
                dialog = builder.create().also {
                    readOnlyMapNotificationDialog = it
                }
                dialog.window?.setSoftInputMode(2)
                dialog.show()
            }
        }
    }

    private fun showLikeDialog() {
        MaterialAlertDialogBuilder(this).apply {
            setTitle(R.string.like_it)
            setMessage(R.string.did_you_like_this_world)
            setCancelable(false)
            setPositiveButton(R.string.like) { dialog: DialogInterface, id: Int ->
                Multiplayer.instance.likeWorld()
                dialog.dismiss()
                showSaveWorldDialog()
            }
            setNegativeButton(R.string.dislike) { dialog: DialogInterface, id: Int ->
                Multiplayer.instance.dislikeWorld()
                dialog.dismiss()
                showSaveWorldDialog()
            }
        }.show()
    }

    private fun sendChatMessage(msg: EditText) {
        val messageText = msg.text.toString()
        if (CHAT_COMMAND_HOME == messageText) {
            game?.blockView?.resetPlayerLocation()
            return
        }
        CoroutineScope(Dispatchers.IO).launch {
            Multiplayer.instance.gameClient?.let { gameClient ->
                if (msg.toString().isNotEmpty()) {
                    gameClient.chat(messageText)
                }
            }
        }
    }

    fun showChatDialog() {
        val msg = EditText(this).apply {
            filters = arrayOf<InputFilter>(LengthFilter(ChatBox.getMaxChatMessageLength()))
        }
        MaterialAlertDialogBuilder(this).apply {
            setTitle(R.string.your_message).setView(msg)
            setPositiveButton(android.R.string.ok) { _: DialogInterface?, _: Int ->
                sendChatMessage(msg)
            }
            setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, _: Int -> dialog.dismiss() }
            create().apply {
                msg.setOnEditorActionListener { _: TextView?, _: Int, event: KeyEvent? ->
                    if (event == null || event.keyCode != 66) {
                        return@setOnEditorActionListener false
                    }
                    sendChatMessage(msg)
                    dismiss()
                    true
                }
            }
        }.show()
    }

    public override fun onPause() {
        super.onPause()
        dismissLoadingDialog()
        gameView?.onPause()
    }

    public override fun onResume() {
        super.onResume()
        gameView?.onResume()
        game?.resetTouches()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (event.repeatCount == 0) {
            gameView?.game?.blockView?.let { blockView ->
                if (keyCode == 4) {
                    showGameMenuDialog()
                } else {
                    blockView.onKeyDown(keyCode, event)
                }
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        gameView?.game?.blockView?.onKeyUp(keyCode, event)
        return true
    }

    private fun hideKeyBoard(v: View) {
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(v.windowToken, 0)
    }

    companion object {
        private const val CHAT_COMMAND_HOME = "/home"
    }
}